addons/product/models/product_template.py

```diff
@@ -938,6 +939,76 @@ def _get_first_possible_combination(self, parent_combination=None, necessary_val
         """
         return next(self._get_possible_combinations(parent_combination, necessary_values), self.env['product.template.attribute.value'])
 
+    def _cartesian_product(self, product_template_attribute_values, parent_combination):
+        """
+        Generate all possible combination for attributes values (aka cartesian product).
+        It is equivalent to itertools.product except it skips invalid partial combinations before they are complete.
+
+        Imagine the cartesian product of 'A', 'CD' and range(1_000_000) and let's say that 'A' and 'C' are incompatible.
+        If you use itertools.product or any normal cartesian product, you'll need to filter out of the final result
+        the 1_000_000 combinations that start with 'A' and 'C' . Instead, This implementation will test if 'A' and 'C' are
+        compatible before even considering range(1_000_000), skip it and and continue with combinations that start
+        with 'A' and 'D'.
+
+        It's necessary for performance reason because filtering out invalid combinations from standard Cartesian product
+        can be extremely slow
+
+        :param product_template_attribute_values: the values we want all the possibles combinations of.
+        One list of values by attribute line
+        :return: a generator of product template values combinations (in the form of lists)
+        """
+        if not product_template_attribute_values:
+            raise StopIteration
+
+        all_exclusions = self._get_own_attribute_exclusions()
+        all_exclusions.update(self._get_parent_attribute_exclusions(parent_combination))
+        # 0 means the value is acceptable, greater than 0 means it's rejected
+        current_exclusions = defaultdict(int)
```

Comment:
Useful comment, but please also explain why it gets incremented and then it gets decremented. Naively I would think that once a value is forbidden it would stay that way.

And just for the sake of completeness also specify that the dict has the ptav as key, even if that was rather easy to guess.

Reply:
```python
# The following dict uses product template attribute values as keys
# 0 means the value is acceptable, greater than 0 means it's rejected
# Bear in mind that several values can reject the same value and the latter can only be included in the
#  considered combination if no value rejects it.
# Each time a value is included in the considered combination, the values it rejects are incremented
# When a value is discarded from the considered combination, the values it rejects are decremented
```